state_machine/applications/system-check/tests/imu_test.py

```diff
@@ -0,0 +1,329 @@
+"""
+Python system check script for PyCubed satellite board
+PyCubed Mini mainboard-v02 for Pocketqube Mission
+IMU Sensor Test
+* Author(s): Yashika Batra
+"""
+
+import time
+from ulab import numpy
+from lib import pycubed as cubesat
+
+
+wait_time = 3
+norm = numpy.linalg.norm
+
+
+def request_imu_data(prompt):
+    """
+    Ask the user to type y/any key to start/cancel the test
+    If y, gather IMU data
+    Accelerometer and Gyroscope readings are collected 10 times
+    in wait_time seconds
+    Magnetometer readings are collected at the start and end
+    of wait_time seconds
+    If any other key, return None for all readings
+    """
+    # ask user whether or not to start the test
+    print(prompt, end="")
+    start_test = input("Type Y to start the test, any key to cancel: ")
+
+    # if user opts to cancel the test, return None, handle in the next function
+    if start_test.lower() != "y":
+        return None, None, None
+
+    # else proceed with the test
+    time.sleep(0.5)
+    print("Collecting IMU data...")
+
+    # collect magnetometer reading at the start of wait time
+    start_mag = numpy.array(cubesat.magnetic())
+
+    # collect total accelerometer and gyroscope readings
+    # do 10 reads over wait_time
+    acc_total = numpy.array([0., 0., 0.])
+    gyro_total = numpy.array([0., 0., 0.])
+    for i in range(10):
+        acc_total += numpy.array(cubesat.acceleration())
+        gyro_total += numpy.array(cubesat.gyro())
+        time.sleep(wait_time / 10)
+
+    # collect magnetometer reading at the end of wait time
+    end_mag = numpy.array(cubesat.magnetic())
+
+    # calculate the average acc and gyro readings
+    acc_final = acc_total / 10
+    gyro_final = gyro_total / 10
+
+    # calculate the change in the total magnetometer reading
+    mag = norm(end_mag - start_mag)
+
+    # return acc, gyro, change in mag
+    print("Data Collection Complete")
+    return acc_final, gyro_final, mag
+
+
+def check_gravity_acc(acc, direction):
+    if direction.lower() == "x":
+        return (abs(abs(acc[0]) - 9.8) < 1 and abs(acc[1]) < 1
+                and abs(acc[2]) < 1)
+    elif direction.lower() == "y":
+        return (abs(acc[0]) < 1 and abs(abs(acc[1]) - 9.8) < 1
+                and abs(acc[2]) < 1)
+    elif direction.lower() == "z":
+        return (abs(acc[0]) < 1 and abs(acc[1]) < 1
+                and abs(abs(acc[2]) - 9.8) < 1)
+
+
+def gravity_imu_test(result_dict):
```

Comment:
This is a really clever test you came up with, but this function shouldn't be nearly 100 lines long (and very confusing). 
Why don't you just ask for three readings (facing up, to the left, and forward) in any order. 
Then just compare them to your axis vectors (+x,-x,+y,-y,+z,-z) by magnitude and angular offset (about 9.8m/s^2, and a couple degrees off). 
If you have three matches, then you passed the test.
Otherwise fail, and give the three vectors that you read as your restult_string.